From 8077d303d9cd8ed2e7e29a94605b1cb6c224cbf9 Mon Sep 17 00:00:00 2001
From: Karthikeyan K <karthikk@habana.ai>
Date: Wed, 09 Mar 2022 03:58:22 +0200
Subject: [PATCH] [SW-56706] boot_desc : update image desc design

Update the image descriptor as per new design.

This patch implements usage of the new fields
1) boot_ptrs_img_sz : Size of the complete image(boot desc header ,
image descriptors, public key, signature, CRC) present at offset
0x60000 of spi. This should be used for dma.
2) offset2array: instead of hardcoded location for offset2array, this
would provide us the flexibility to extend boot info header in future.

The is in preparation to port boot info onto sram, and then use the
new fields.

Change-Id: I9e20d7c86c97bbf406ca8322121905cc25203374
---

diff --git a/common/boot_desc.c b/common/boot_desc.c
index 15aa11b..59a88f0 100644
--- a/common/boot_desc.c
+++ b/common/boot_desc.c
@@ -56,10 +56,11 @@
 
 int boot_desc_init(enum hl_fw_boot_ptr_imgs img)
 {
-	uint32_t i, boot_ptrs_count, img_ptrs_size;
+	uint32_t i, img_ptrs_size;
 	uint32_t *p_offset = NULL;
 	enum arcboot_rom_progress dma_sts_start = PROGRESS_INIT_SPI_BOOT_DESC_DMA_START;
 	enum arcboot_rom_progress dma_sts_end = PROGRESS_INIT_SPI_BOOT_DESC_DMA_FINISH;
+	enum dma_error err;
 
 	if (img >= HL_FW_BOOT_PTR_MAX_IMG)
 		return 0;
@@ -70,19 +71,7 @@
 	if (p_img_ptr_ctx->boot_desc_flag)
 		return p_boot_ptr->count;
 
-	boot_ptrs_count = spi_flash_readl(p_img_ptr_ctx->img_ptr_spi_offset
-			+ HL_BOOT_PTS_COUNT_FLASH_OFFSET);
-
-	if (boot_ptrs_count >= p_img_ptr_ctx->max_imgs)
-		return 0;
-
-	img_ptrs_size = BOOT_PTRS_SZ + ((sizeof(uint32_t) +
-			IMG_DESC_SZ) * boot_ptrs_count);
-
-	/* Align image size to 16 bytes */
-	img_ptrs_size = ALIGN_VAL(img_ptrs_size, 16);
-	/* Add 4 byte CRC to aligned image size for DMA */
-	img_ptrs_size += CRC32_SIZE;
+	img_ptrs_size = spi_flash_readl(p_img_ptr_ctx->img_ptr_spi_offset);
 
 	if (img == HL_FW_BOOT_PTR_EROM) {
 		dma_sts_start = PROGRESS_ARCBOOT_EROM_BOOT_DESC_DMA_START;
@@ -92,20 +81,30 @@
 	arcboot_rom_status_update(dma_sts_start);
 
 	/* Fetch boot desc using DMA into dest offset */
-	dma_copy_image(img_ptrs_size, p_img_ptr_ctx->img_ptr_spi_offset,
-			HL_BOOT_PTR_DMA_DEST, SPI_DMA_DST_HBW);
-	if (dma_wait_finish() == DMA_CRC_ERROR)
+	err = dma_copy_image_v2(img_ptrs_size, p_img_ptr_ctx->img_ptr_spi_offset,
+			HL_BOOT_PTR_DMA_DEST, DMA_DEST_HBW, false);
+	if (err != DMA_SUCCESS) {
+		if (err == DMA_INVALID_CRC_ERROR)
+			arcboot_rom_status_update(PROGRESS_INIT_SPI_BOOT_DESC_CRC_FAIL);
+		else if (err == DMA_UNALIGNED_DEST_ADDR)
+			arcboot_rom_status_update(PROGRESS_INIT_SPI_BOOT_DESC_UNALIGNED_DEST);
+		else if (err == DMA_UNALIGNED_IMG_SIZE)
+			arcboot_rom_status_update(PROGRESS_INIT_SPI_BOOT_DESC_UNALIGNED_IMG_SIZE);
+		else
+			arcboot_rom_status_update(PROGRESS_INIT_SPI_BOOT_DESC_DMA_ERROR);
 		return 0;
+	}
 
 	arcboot_rom_status_update(dma_sts_end);
 
-	p_offset = (uint32_t *)(HL_BOOT_PTR_DEST_ARC + BOOT_PTRS_SZ);
 	memcpy(p_boot_ptr,
 		(uint8_t *)(HL_BOOT_PTR_DEST_ARC), BOOT_PTRS_SZ);
 
 	if (p_boot_ptr->magic.val != HL_FW_BOOT_MAGIC)
 		return 0;
 
+	p_offset = (uint32_t *)(HL_BOOT_PTR_DEST_ARC + p_boot_ptr->offset2array);
+
 	/* Read all descriptors */
 	for (i = 0; i < p_boot_ptr->count; i++) {
 		memcpy(&p_img_desc[i],
diff --git a/include/arcboot.h b/include/arcboot.h
index 0267a39..1df7892 100644
--- a/include/arcboot.h
+++ b/include/arcboot.h
@@ -96,66 +96,70 @@
 	PROGRESS_INIT_SP_SRAM_NOT_COMPLETED = 1,
 	PROGRESS_INIT_ARC1_ICCM_INIT_FAILED = 2,
 	PROGRESS_INIT_SPI_BOOT_DESC_DMA_START = 3,
-	PROGRESS_INIT_SPI_BOOT_DESC_FAIL = 4,
-	PROGRESS_INIT_SPI_BOOT_DESC_DMA_FINISH = 5,
-	PROGRESS_INIT_EFUSE_SPI = 6,
-	PROGRESS_REVOKE_LOC = 7,
-	PROGRESS_REVOKE_LOC_0 = 8,
-	PROGRESS_REVOKE_LOC_1 = 9,
-	PROGRESS_KEY_LOC = 10,
-	PROGRESS_KEY_LOC_0 = 11,
-	PROGRESS_KEY_LOC_1 = 12,
-	PROGRESS_VER_LOC = 13,
-	PROGRESS_VER_LOC_0 = 14,
-	PROGRESS_VER_LOC_1 = 15,
-	PROGRESS_BRINGUP_PCI_PRI = 16,
-	PROGRESS_BRINGUP_PCI_PRI_TOO_BIG = 17,
-	PROGRESS_BRINGUP_PCI_PRI_CRC_FAIL = 18,
-	PROGRESS_BRINGUP_PCI_PRI_UNALIGNED_DEST_ADDR = 19,
-	PROGRESS_BRINGUP_PCI_PRI_UNALIGNED_IMG_SIZE = 20,
-	PROGRESS_BRINGUP_PCI_PRI_DMA = 21,
-	PROGRESS_BRINGUP_PCI_PRI_KEY_REVOKED = 22,
-	PROGRESS_BRINGUP_PCI_PRI_KEY_AUTH_FAIL = 23,
-	PROGRESS_BRINGUP_PCI_PRI_SECURITY = 24,
-	PROGRESS_BRINGUP_PCI_SEC = 25,
-	PROGRESS_BRINGUP_PCI_SEC_TOO_BIG = 26,
-	PROGRESS_BRINGUP_PCI_SEC_CRC_FAIL = 27,
-	PROGRESS_BRINGUP_PCI_SEC_UNALIGNED_DEST_ADDR = 28,
-	PROGRESS_BRINGUP_PCI_SEC_UNALIGNED_IMG_SIZE = 29,
-	PROGRESS_BRINGUP_PCI_SEC_DMA = 30,
-	PROGRESS_BRINGUP_PCI_SEC_KEY_REVOKED = 31,
-	PROGRESS_BRINGUP_PCI_SEC_KEY_AUTH_FAIL = 32,
-	PROGRESS_BRINGUP_PCI_SEC_SECURITY = 33,
-	PROGRESS_PCIE_IN_RECOVERY = 34,
-	PROGRESS_PPBOOT_START = 35,
-	PROGRESS_PPBOOT_FINISH = 36,
-	PROGRESS_BRINGUP_RAM_PRI = 37,
-	PROGRESS_BRINGUP_RAM_PRI_DMA = 38,
-	PROGRESS_BRINGUP_RAM_PRI_SECURITY = 39,
-	PROGRESS_BRINGUP_RAM_SEC = 40,
-	PROGRESS_BRINGUP_RAM_SEC_DMA = 41,
-	PROGRESS_BRINGUP_RAM_SEC_SECURITY = 42,
-	PROGRESS_ARCBOOT_RAM_START = 43,
-	PROGRESS_ARCBOOT_RECOVERY = 44,
-	PROGRESS_ARCBOOT_RECOVERY_DMA = 45,
-	PROGRESS_ARCBOOT_EROM_BOOT_DESC_ERROR = 46,
-	PROGRESS_ARCBOOT_EROM_BOOT_DESC_DMA_START = 47,
-	PROGRESS_ARCBOOT_EROM_BOOT_DESC_DMA_FINISH = 48,
-	PROGRESS_ARCBOOT_EROM_PCIE = 49,
-	PROGRESS_ARCBOOT_EROM_PCIE_DMA_START = 50,
-	PROGRESS_ARCBOOT_EROM_PCIE_DMA_FINISH = 51,
-	PROGRESS_ARCBOOT_EROM_PCIE_START = 52,
-	PROGRESS_ARCBOOT_EROM_PCIE_FINISH = 53,
-	PROGRESS_ARCBOOT_EROM_RECOVERY = 54,
-	PROGRESS_ARCBOOT_EROM_RECOVERY_DMA_START = 55,
-	PROGRESS_ARCBOOT_EROM_RECOVERY_DMA_FINISH = 56,
-	PROGRESS_ARCBOOT_EROM_RECOVERY_START = 57,
-	PROGRESS_ARCBOOT_EROM_RECOVERY_FINISH = 58,
-	PROGRESS_ARCBOOT_EROM_TPM = 59,
-	PROGRESS_ARCBOOT_EROM_TPM_DMA_START = 60,
-	PROGRESS_ARCBOOT_EROM_TPM_DMA_FINISH = 61,
-	PROGRESS_ARCBOOT_EROM_TPM_START = 62,
-	PROGRESS_ARCBOOT_EROM_TPM_FINISH = 63,
+	PROGRESS_INIT_SPI_BOOT_DESC_CRC_FAIL = 4,
+	PROGRESS_INIT_SPI_BOOT_DESC_UNALIGNED_DEST = 5,
+	PROGRESS_INIT_SPI_BOOT_DESC_UNALIGNED_IMG_SIZE = 6,
+	PROGRESS_INIT_SPI_BOOT_DESC_DMA_ERROR = 7,
+	PROGRESS_INIT_SPI_BOOT_DESC_FAIL = 8,
+	PROGRESS_INIT_SPI_BOOT_DESC_DMA_FINISH = 9,
+	PROGRESS_INIT_EFUSE_SPI = 10,
+	PROGRESS_REVOKE_LOC = 11,
+	PROGRESS_REVOKE_LOC_0 = 12,
+	PROGRESS_REVOKE_LOC_1 = 13,
+	PROGRESS_KEY_LOC = 14,
+	PROGRESS_KEY_LOC_0 = 15,
+	PROGRESS_KEY_LOC_1 = 16,
+	PROGRESS_VER_LOC = 17,
+	PROGRESS_VER_LOC_0 = 18,
+	PROGRESS_VER_LOC_1 = 19,
+	PROGRESS_BRINGUP_PCI_PRI = 20,
+	PROGRESS_BRINGUP_PCI_PRI_TOO_BIG = 21,
+	PROGRESS_BRINGUP_PCI_PRI_CRC_FAIL = 22,
+	PROGRESS_BRINGUP_PCI_PRI_UNALIGNED_DEST_ADDR = 23,
+	PROGRESS_BRINGUP_PCI_PRI_UNALIGNED_IMG_SIZE = 24,
+	PROGRESS_BRINGUP_PCI_PRI_DMA = 25,
+	PROGRESS_BRINGUP_PCI_PRI_KEY_REVOKED = 26,
+	PROGRESS_BRINGUP_PCI_PRI_KEY_AUTH_FAIL = 27,
+	PROGRESS_BRINGUP_PCI_PRI_SECURITY = 28,
+	PROGRESS_BRINGUP_PCI_SEC = 29,
+	PROGRESS_BRINGUP_PCI_SEC_TOO_BIG = 30,
+	PROGRESS_BRINGUP_PCI_SEC_CRC_FAIL = 31,
+	PROGRESS_BRINGUP_PCI_SEC_UNALIGNED_DEST_ADDR = 32,
+	PROGRESS_BRINGUP_PCI_SEC_UNALIGNED_IMG_SIZE = 33,
+	PROGRESS_BRINGUP_PCI_SEC_DMA = 34,
+	PROGRESS_BRINGUP_PCI_SEC_KEY_REVOKED = 35,
+	PROGRESS_BRINGUP_PCI_SEC_KEY_AUTH_FAIL = 36,
+	PROGRESS_BRINGUP_PCI_SEC_SECURITY = 37,
+	PROGRESS_PCIE_IN_RECOVERY = 38,
+	PROGRESS_PPBOOT_START = 39,
+	PROGRESS_PPBOOT_FINISH = 40,
+	PROGRESS_BRINGUP_RAM_PRI = 41,
+	PROGRESS_BRINGUP_RAM_PRI_DMA = 42,
+	PROGRESS_BRINGUP_RAM_PRI_SECURITY = 43,
+	PROGRESS_BRINGUP_RAM_SEC = 44,
+	PROGRESS_BRINGUP_RAM_SEC_DMA = 45,
+	PROGRESS_BRINGUP_RAM_SEC_SECURITY = 46,
+	PROGRESS_ARCBOOT_RAM_START = 47,
+	PROGRESS_ARCBOOT_RECOVERY = 48,
+	PROGRESS_ARCBOOT_RECOVERY_DMA = 49,
+	PROGRESS_ARCBOOT_EROM_BOOT_DESC_ERROR = 50,
+	PROGRESS_ARCBOOT_EROM_BOOT_DESC_DMA_START = 51,
+	PROGRESS_ARCBOOT_EROM_BOOT_DESC_DMA_FINISH = 52,
+	PROGRESS_ARCBOOT_EROM_PCIE = 53,
+	PROGRESS_ARCBOOT_EROM_PCIE_DMA_START = 54,
+	PROGRESS_ARCBOOT_EROM_PCIE_DMA_FINISH = 55,
+	PROGRESS_ARCBOOT_EROM_PCIE_START = 56,
+	PROGRESS_ARCBOOT_EROM_PCIE_FINISH = 57,
+	PROGRESS_ARCBOOT_EROM_RECOVERY = 58,
+	PROGRESS_ARCBOOT_EROM_RECOVERY_DMA_START = 59,
+	PROGRESS_ARCBOOT_EROM_RECOVERY_DMA_FINISH = 60,
+	PROGRESS_ARCBOOT_EROM_RECOVERY_START = 61,
+	PROGRESS_ARCBOOT_EROM_RECOVERY_FINISH = 62,
+	PROGRESS_ARCBOOT_EROM_TPM = 63,
+	PROGRESS_ARCBOOT_EROM_TPM_DMA_START = 64,
+	PROGRESS_ARCBOOT_EROM_TPM_DMA_FINISH = 65,
+	PROGRESS_ARCBOOT_EROM_TPM_START = 66,
+	PROGRESS_ARCBOOT_EROM_TPM_FINISH = 67,
 };
 
 enum security_location {
